packages/ember-metal/lib/meta_listeners.js

```diff
@@ -20,6 +22,24 @@ export var protoMethods = {
     if (!this._listeners) {
       this._listeners = [];
     }
+
+    runInDebug(() => {
+      // Assert that an observer is only used once.
+      let listeners = this._listeners;
+      let matchFound = false;
+
+      for (let i = listeners.length - 4; i >= 0; i -= 4) {
+        if (listeners[i+1] === target &&
+            listeners[i+2] === method &&
+            listeners[i] === eventName) {
+          matchFound = true;
+          break;
+        }
+      }
+
+      assert(`Tried to add a duplicate listener for ${eventName}`, !matchFound);
```

Comment:
It might be nice to have the object's toString here in the message, to make it easier to track down the source. Would that work?